### DISCUSSION TITLE
Lesson 9- Where Exactly does a player pays the Entrance Fee to participate in the Raffle

### DISCUSSION THREAD

**Discussion by debay1010** [Cyfrin/foundry-full-course-cu#895] [Q&A]
Looking at the function enterRaffle(), can somebody help me point out the exact place where a player pays the entrance fee to participate in the Raffle.  The understanding I had uses mapping, but this project did not use mapping. So I am at loss when I tried to figure out where the player pays the entrance fee.

This code, to my understanding adds the player's address to the array:
```
s_players.push(payable(msg.sender));
```
And this only checks the amount if it is up to the entrance fee:
```
  if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEntranceFee();
```
So where did the player pay? Thanks.

Function enterRaffle()
---------------------
```code

function enterRaffle() external payable {
        if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEntranceFee();
        }
        if (s_raffleState != RaffleState.OPEN) {
            revert Raffle__RaffleNotOpen();
        }
        s_players.push(payable(msg.sender));
        emit EnteredRaffle(msg.sender);
    }
```

**Comment by 0xusmanf** [comment]
> Looking at the function enterRaffle(), can somebody help me point out the exact place where a player pays the entrance fee to participate in the Raffle. The understanding I had uses mapping, but this project did not use mapping. So I am at loss when I tried to figure out where the player pays the entrance fee.

The FundMe project uses mapping to keep track of funding (how much a user has funded), mapping is a data structure just like array, it does not have anything to do with adding balance to the contract. Users still would have been able to fund the contract even if it did not use the mapping.

In both cases (FundMe and this) `msg.value` sent by users is added to the `address(this).balance` variable which every contract has by default.

We use arrays and mappings to only keep the record of how much a user has funded or how many users have entered the raffle because otherwise, we wouldn't know by just `address(this).balance` as it will only give you the total ether balance of the contract.

**Reply by debay1010** [reply]
@Usmanfarooq91, this statement is very instructive

"We use arrays and mappings to only keep the record of how much a user has funded or how many users have entered the raffle because otherwise, we wouldn't know by just address(this).balance as it will only give you the total ether balance of the contract."

Thanks.

**Comment by shrutic11** [comment]
if you notice, the function has a keyword "payable". Any function in a contract that is marked payable can receive funds for that particular smart contract.

So when a user is executing the following command:
```solidity
raffleLottery.enterRaffle{value: ENTRANCE_FEE}();
```

He is using the raffle contract object -> raffleLottery
to call the payable function --> enterRaffle
with the value {ENTRANCE_FEE}
which will send eth FROM his test account/wallet, depending on from where he is calling the function TO the RaffleLottery.sol smart contract.

For example, here I am trying to enter the raffle as a test user:

```solidity
function testPlayerCanEnterRaffleWithEntranceFee() public {
        vm.prank(TEST_PLAYER);

        vm.expectEmit(address(raffleLottery));
        emit NewParticipant(TEST_PLAYER);
        raffleLottery.enterRaffle{value: ENTRANCE_FEE}();

        assertEq(raffleLottery.getParticipants()[0], TEST_PLAYER);
    }
 ```

    
```
forge test --mt testPlayerCanEnterRaffleWithEntranceFee
```

**Reply by debay1010** [reply]
@shrutic11, Thanks for your reply. 

Now suppose a function in a contract is marked as payable and does not contain a statement like
```
if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEntranceFee();
 ```
 And when the function is being called, no ether is sent, what will happen? Thanks.

**Reply by shrutic11** [reply]
The transaction will succeed with 0 wei to transfer and will still coast you gas for the operation. 

<img width="807" alt="Screenshot 2023-09-22 at 12 35 45 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/29403554/33dbb879-2f7b-4984-97e0-646a37283841">


In case of your code:
```
function enterRaffle() external payable {
        if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEntranceFee();
        }
        if (s_raffleState != RaffleState.OPEN) {
            revert Raffle__RaffleNotOpen();
        }
        s_players.push(payable(msg.sender));
        emit EnteredRaffle(msg.sender);
    }
```
If we remove the entranceFee check, the function will execute the next line, which is Raffle State check, and if raffle state is open, it will move to next line and add the caller of the function i.e msg.sender in the s_players array, even though he didn't pay any entrance fee (because we removed the entrance fee value check).
